src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/Touch/GestureManager.cs

```diff
@@ -0,0 +1,789 @@
+		{
+			if (isToggled.HasValue)
+			{
+				Console.WriteLine($"Touch state: {status}");
+				var r = (status == TouchStatus.Started && isToggled.Value) ||
+					(status != TouchStatus.Started && !isToggled.Value);
+				state = r
+					? TouchState.Normal
+					: TouchState.Pressed;
+			}
+
+			await RunAnimationTask(sender, state, hoverState, animationTokenSource.Token).ConfigureAwait(false);
+			return;
+		}
+		do
+		{
+			var rippleState = isToggled.HasValue && isToggled.Value
+				? TouchState.Normal
+				: TouchState.Pressed;
+
+			await RunAnimationTask(sender, rippleState, hoverState, animationTokenSource.Token);
+			if (token.IsCancellationRequested)
+			{
+				return;
+			}
+
+			rippleState = isToggled.HasValue && isToggled.Value
+				? TouchState.Pressed
+				: TouchState.Normal;
+
+			await RunAnimationTask(sender, rippleState, hoverState, animationTokenSource.Token);
+			if (token.IsCancellationRequested)
+			{
+				return;
+			}
+		}
+		while (--pulseCount != 0);
+	}
+
+	internal void HandleLongPress(TouchBehavior sender)
+	{
+		if (sender.State == TouchState.Normal)
+		{
+			longPressTokenSource?.Cancel();
+			longPressTokenSource?.Dispose();
+			longPressTokenSource = null;
+			return;
+		}
+
+		if (sender.LongPressCommand is null || sender.InteractionStatus == TouchInteractionStatus.Completed)
+		{
+			return;
+		}
+
+		longPressTokenSource = new CancellationTokenSource();
+		Task.Delay(sender.LongPressDuration, longPressTokenSource.Token).ContinueWith(t =>
+		{
+			if (t.IsFaulted && t.Exception != null)
+			{
+				throw t.Exception;
+			}
+
+			if (t.IsCanceled)
+			{
+				return;
+			}
+
+			var longPressAction = new Action(() =>
+			{
+				sender.HandleUserInteraction(TouchInteractionStatus.Completed);
+				sender.RaiseLongPressCompleted();
+			});
+			
+			if (sender.Dispatcher.IsDispatchRequired)
+			{
+				sender.Dispatcher.Dispatch(longPressAction);
+			}
+			else
+			{
+				longPressAction.Invoke();
+			}
+		});
+	}
+
+	internal void SetCustomAnimationTask(Func<TouchBehavior, TouchState, HoverState, int, Easing?, CancellationToken, Task>? animationTaskFactory)
+		=> this.animationTaskFactory = animationTaskFactory;
+
+	internal void Reset()
+	{
+		SetCustomAnimationTask(null);
+		defaultBackgroundColor = default;
+	}
+
+	internal void OnTapped(TouchBehavior sender)
+	{
+		if (!sender.CanExecute || (sender.LongPressCommand != null && sender.InteractionStatus == TouchInteractionStatus.Completed))
+		{
+			return;
+		}
+
+		if (DeviceInfo.Platform == DevicePlatform.Android)
+		{
+			HandleCollectionViewSelection(sender);
+		}
+
+		if (sender.Element is IButtonController button)
+		{
+			button.SendClicked();
+		}
+
+		sender.RaiseCompleted();
+	}
+
+	void HandleCollectionViewSelection(TouchBehavior sender)
+	{
+		CollectionView? parent = null;
+		VisualElement? result = null;
+		if (!sender.Element?.TryFindParentElementWithParentOfType(out result, out parent) ?? true)
```

Comment:
Note for the future, on `TryFindParentElementWithParentOfType` add attributes to guide roslyn that `out` can't be null if the return is `true`